src/stream/src/executor/managed_state/join/join_row_set.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::btree_map::OccupiedError as BTreeMapOccupiedError;
+use std::collections::BTreeMap;
+use std::fmt::Debug;
+use std::mem;
+
+use auto_enums::auto_enum;
+use enum_as_inner::EnumAsInner;
+
+const MAX_VEC_SIZE: usize = 4;
+
+#[derive(Debug, EnumAsInner)]
+pub enum JoinRowSet<K, V> {
+    BTree(BTreeMap<K, V>),
+    Vec(Vec<(K, V)>),
+}
+
+impl<K, V> Default for JoinRowSet<K, V> {
+    fn default() -> Self {
+        Self::Vec(Vec::new())
+    }
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct VecOccupiedError<'a, K, V> {
+    key: &'a K,
+    old_value: &'a V,
+    new_value: V,
+}
+
+#[derive(Debug)]
+pub enum JoinRowSetOccupiedError<'a, K: Ord, V> {
+    BTree(BTreeMapOccupiedError<'a, K, V>),
+    Vec(VecOccupiedError<'a, K, V>),
+}
+
+impl<K: Ord + Debug, V: Debug> JoinRowSet<K, V> {
+    pub fn try_insert(
+        &mut self,
+        key: K,
+        value: V,
+    ) -> Result<&'_ mut V, JoinRowSetOccupiedError<'_, K, V>> {
+        if let Self::Vec(inner) = self && inner.len() >= MAX_VEC_SIZE{
+            let btree = BTreeMap::from_iter(inner.drain(..));
+            mem::swap(self, &mut Self::BTree(btree));
+        }
+
+        match self {
+            Self::BTree(inner) => inner
+                .try_insert(key, value)
+                .map_err(JoinRowSetOccupiedError::BTree),
+            Self::Vec(inner) => {
+                if let Some(pos) = inner.iter().position(|elem| elem.0 == key) {
+                    Err(JoinRowSetOccupiedError::Vec(VecOccupiedError {
+                        key: &inner[pos].0,
+                        old_value: &inner[pos].1,
+                        new_value: value,
+                    }))
+                } else {
+                    inner.push((key, value));
```

Comment:
What about reserving exactly `MAX_VEC_SIZE` capacity on each insertion? Otherwise it seems possible that most capacity of the last allocation is wasted.

Reply:
capacity reserved is not trivial, I'll try optimize